ts/WoltLabSuite/Core/Api.ts

```diff
@@ -0,0 +1,294 @@
+    this.className = className;
+  }
+
+  static dboAction(actionName: string, className: string): Api {
+    return new Api(actionName, className);
+  }
+
+  failure(failure: CallbackFailure): this {
+    this._failure = failure;
+
+    return this;
+  }
+
+  getAbortController(): AbortController {
+    if (this._signal === undefined) {
+      this._signal = new AbortController();
+    }
+
+    return this._signal;
+  }
+
+  objectIds(objectIds: number[]): this {
+    this._objectIDs = objectIds;
+
+    return this;
+  }
+
+  payload(payload: Payload): this {
+    this._payload = payload;
+
+    return this;
+  }
+
+  disableLoadingIndicator(): this {
+    this._showLoadingIndicator = false;
+
+    return this;
+  }
+
+  async dispatch(): Promise<unknown> {
+    const url = window.WSC_API_URL + "index.php?ajax-proxy/&t=" + Core.getXsrfToken();
+
+    const body: RequestBody = {
+      actionName: this.actionName,
+      className: this.className,
+    };
+    if (this._objectIDs) {
+      body.objectIDs = this._objectIDs;
+    }
+    if (this._payload) {
+      body.parameters = this._payload;
+    }
+
+    const init: RequestInit = {
+      method: "POST",
+      headers: {
+        "Content-Type": "application/x-www-form-urlencoded; charset=UTF-8",
+        "X-Requested-With": "XMLHttpRequest",
+        "X-XSRF-TOKEN": Core.getXsrfToken(),
+      },
+      body: Core.serialize(body),
+      mode: "same-origin",
+      credentials: "same-origin",
+      cache: "no-store",
+      redirect: "error",
+    };
+
+    if (this._signal) {
+      init.signal = this._signal.signal;
+    }
+
+    // Use a local copy to isolte the behavior in case of changes before
+    // the request handling has completed.
+    const showLoadingIndicator = this._showLoadingIndicator;
+    if (showLoadingIndicator) {
+      AjaxStatus.show();
+    }
+
+    try {
+      const response = await fetch(url, init);
+
+      if (!response.ok) {
+        const result = this.handleError(ErrorCode.STATUS_NOT_OK, response);
+
+        return Promise.reject(result);
+      }
+
+      const contentType = response.headers.get("content-type");
+      if (!contentType || !contentType.includes("application/json")) {
+        const result = this.handleError(ErrorCode.EXPECTED_JSON, response);
+
+        return Promise.reject(result);
+      }
+
+      let json: ResponseData;
+      try {
+        json = await response.json();
+      } catch (e) {
+        const result = this.handleError(ErrorCode.INVALID_JSON, response);
+
+        return Promise.reject(result);
+      }
+
+      // This is an explicit wait for the event loop to execute the
+      // callee function before we execute additional tasks.
+      return Promise.resolve(json.returnValues).then(async (result) => {
+        if (json.forceBackgroundQueuePerform) {
+          const backgroundQueue = await import("./BackgroundQueue");
+          backgroundQueue.invoke();
+        }
+        return result;
+      });
```

Comment:
Not sure whether I would do this with a `.then()`, because that will require the loading of the background queue to succeed for the AJAX request to succeed. Instead:

```
if (json.forceBackgroundQueuePerform) {
 // Fire and forget.
 void import("./BackgroundQueue").then((q) => q.invoke());
}

// an explicit Promise.resolve() is no longer required.
return json.returnValues;
```